### PR TITLE
feat: put pbj consensus data into gossip event

### LINKED ISSUE #13308: Put PBJ consensus data into GossipEvent

### PR DESCRIPTION
**Pull request by lpetrovic05**
closes #13308

### CODE REVIEW

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/GossipEvent.java [hunk 1]
```diff
@@ -332,7 +386,8 @@ public boolean equals(final Object o) {
         }
 
         final GossipEvent that = (GossipEvent) o;
-        return Objects.equals(getHashedData(), that.getHashedData());
+        return Objects.equals(getHashedData(), that.getHashedData())
+                && Objects.equals(consensusData, that.consensusData);
```
Comment (cody-littley) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/GossipEvent.java:390]: Not something for this pr.

In production, the only way we should ever compare events is by their hashes. `equals()` and `hashCode()` methods are essentially unit test code that live in production classes. I advocate that we remove them, or perhaps even have them throw an unsupported operation exception. 

We currently only use them in unit tests. But for unit tests, we can implement static comparison methods.

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/recovery/internal/EventStreamRoundLowerBound.java [hunk 2]
```diff
@@ -43,8 +43,8 @@ public EventStreamRoundLowerBound(final long round) {
      * {@inheritDoc}
      */
     @Override
-    public int compareTo(ConsensusData consensusData) {
-        return Long.compare(consensusData.getRoundReceived(), round);
+    public int compareTo(DetailedConsensusEvent consensusEvent) {
```
Comment (cody-littley) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/recovery/internal/EventStreamRoundLowerBound.java:46]:
```suggestion
    public int compareTo(@NonNull final DetailedConsensusEvent consensusEvent) {
```

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/recovery/internal/EventStreamTimestampLowerBound.java [hunk 3]
```diff
@@ -42,7 +42,7 @@ public EventStreamTimestampLowerBound(@NonNull final Instant timestamp) {
      * {@inheritDoc}
      */
     @Override
-    public int compareTo(ConsensusData consensusData) {
-        return consensusData.getConsensusTimestamp().compareTo(timestamp);
+    public int compareTo(DetailedConsensusEvent consensusEvent) {
```
Comment (cody-littley) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/recovery/internal/EventStreamTimestampLowerBound.java:45]:
```suggestion
    public int compareTo(@NonNull final DetailedConsensusEvent consensusEvent) {
```

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/system/events/DetailedConsensusEvent.java [hunk 4]
```diff
@@ -25,54 +27,86 @@
...
      * Creates an empty instance
      */
     public DetailedConsensusEvent() {}
 
+    /**
+     * Create a new instance.
+     *
+     * @param event the event to copy the data from
+     */
+    public DetailedConsensusEvent(final EventImpl event) {
```
Comment (cody-littley) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/system/events/DetailedConsensusEvent.java:65]:
```suggestion
    public DetailedConsensusEvent(@NonNull final EventImpl event) {
```

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/system/events/DetailedConsensusEvent.java [hunk 5]
```diff
@@ -25,54 +27,86 @@
...
+     * @param lastInRoundReceived true if this event is the last in consensus order of all those with the same received
+     *                            round
      */
-    public DetailedConsensusEvent(final GossipEvent gossipEvent, final ConsensusData consensusData) {
+    public DetailedConsensusEvent(
+            final GossipEvent gossipEvent, final long roundReceived, final boolean lastInRoundReceived) {
         this.gossipEvent = gossipEvent;
-        this.consensusData = consensusData;
+        this.roundReceived = roundReceived;
+        this.lastInRoundReceived = lastInRoundReceived;
```
Comment (cody-littley) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/system/events/DetailedConsensusEvent.java:83]: nullity annotations + checks

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/system/events/DetailedConsensusEvent.java [hunk 6]
```diff
@@ -25,54 +27,86 @@
...
     }
 
     public static void serialize(
-            final SerializableDataOutputStream out, final GossipEvent gossipEvent, final ConsensusData consensusData)
+            final SerializableDataOutputStream out,
+            final GossipEvent gossipEvent,
+            final long roundReceived,
+            final boolean lastInRoundReceived)
             throws IOException {
```
Comment (cody-littley) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/system/events/DetailedConsensusEvent.java:91]: nullity annotations

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/system/events/DetailedConsensusEvent.java [hunk 7]
```diff
@@ -82,7 +116,21 @@ public void serialize(final SerializableDataOutputStream out) throws IOException
...
+        lastInRoundReceived = in.readBoolean();
+        final Instant consensusTimestamp = in.readInstant();
+        roundReceived = in.readLong();
+        final long consensusOrder = in.readLong();
+
+        final EventConsensusData eventConsensusData = EventConsensusData.newBuilder()
+                .consensusTimestamp(HapiUtils.asTimestamp(consensusTimestamp))
+                .consensusOrder(consensusOrder)
+                .build();
+        gossipEvent.setConsensusData(eventConsensusData);
```
Comment (cody-littley) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/system/events/DetailedConsensusEvent.java:133]: Does the byte format for this change? Not a huge deal if it does... but if so we should probably get signoff that it's ok to change byte format in CES without HIP.

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/ConsensusImpl.java [hunk 8]
```diff
@@ -791,11 +794,9 @@ private void setIsConsensusTrue(@NonNull final EventImpl event, final long recei
         final List<Instant> times = event.getRecTimes();
 
         // take middle. If there are 2 middle (even length) then use the 2nd (max) of them
-        event.setConsensusTimestamp(times.get(times.size() / 2));
+        event.setPreliminaryConsensusTimestamp(times.get(times.size() / 2));
 
         event.setReachedConsTimestamp(Instant.now()); // used for statistics
-
-        consensusMetrics.consensusReached(event);
     }
```
Comment (alittley) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/ConsensusImpl.java:800]: having removed `consensusMetrics`, this method `setIsConsensusTrue` may now be `static`

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/EventMetadata.java [hunk 9]
```diff
@@ -47,6 +47,11 @@ public class EventMetadata implements Clearable {
     private boolean isJudge;
     /** is this part of the consensus order yet? */
     private boolean isConsensus;
+    /**
+     * a field used to store consensus time while it is still not finalized. depending on the phase of consensus
+     * calculation, this filed may or may not store the final consensus time.
```
Comment (alittley) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/EventMetadata.java:52]: `filed` -> `field`

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/EventUtils.java [hunk 10]
```diff
@@ -54,4 +56,38 @@ public static Instant calculateNewEventCreationTime(
             return now;
         }
     }
+
+    /**
+     * Returns the timestamp of the transaction with given index in this event
+     *
+     * @param transactionIndex index of the transaction in this event
```
Comment (alittley) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/EventUtils.java:63]: missing `@param` doc for `event`

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/EventUtils.java [hunk 11]
```diff
@@ -54,4 +56,38 @@ public static Instant calculateNewEventCreationTime(
...
+     * @return timestamp of the given index transaction
+     */
+    public static @NonNull Instant getTransactionTime(@NonNull final GossipEvent event, final int transactionIndex) {
+        if (event.getConsensusTimestamp() == null) {
+            throw new IllegalArgumentException("Event is not a consensus event");
+        }
+        if (transactionIndex >= event.getPayloadCount()) {
+            throw new IllegalArgumentException("Event does not have a transaction with index:" + transactionIndex);
```
Comment (alittley) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/EventUtils.java:71]: missing trailing space after `:` in exception string

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/EventUtils.java [hunk 12]
```diff
@@ -54,4 +56,38 @@ public static Instant calculateNewEventCreationTime(
...
+        return event.getConsensusTimestamp().plusNanos(transactionIndex * MIN_TRANS_TIMESTAMP_INCR_NANOS);
+    }
+
+    /**
+     * Returns the timestamp of the last transaction in this event. If this event has no transaction, then the timestamp
+     * of the event will be returned
+     *
+     * @return timestamp of the last transaction
+     */
+    public static @NonNull Instant getLastTransTime(@NonNull final GossipEvent event) {
```
Comment (alittley) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/EventUtils.java:82]: missing `@param` doc for `event`

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/GossipEvent.java (hunk 1 above)
Comment (hendrikebbers) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/GossipEvent.java:390]: > throw an unsupported operation exception.
That would not allow to store that data types in any hash collection.

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/internal/EventImpl.java [hunk 13]
```diff
@@ -101,7 +96,15 @@ public EventImpl(@NonNull final GossipEvent gossipEvent) {
         this(gossipEvent, new ConsensusData(), null, null);
     }
 
-    public EventImpl(
+    /**
+     * Create an instance based on the given {@link DetailedConsensusEvent}
+     * @param detailedConsensusEvent the detailed consensus event to build from
+     */
+    public EventImpl(final DetailedConsensusEvent detailedConsensusEvent) {
```
Comment (hendrikebbers) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/internal/EventImpl.java:103]: null annotation

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/system/events/DetailedConsensusEvent.java [hunk 14]
```diff
@@ -25,54 +27,86 @@
...
+            final SerializableDataOutputStream out,
+            final GossipEvent gossipEvent,
+            final long roundReceived,
+            final boolean lastInRoundReceived)
             throws IOException {
         gossipEvent.serialize(out);
-        out.writeSerializable(consensusData, false);
+
+        out.writeInt(ConsensusData.CLASS_VERSION);
+        out.writeLong(UNDEFINED); // ConsensusData.generation
```
Comment (hendrikebbers) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/system/events/DetailedConsensusEvent.java:95]: Those comments are not helpfull. Maybe you can add 3 more words to understand the meaning.

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/GossipEvent.java (hunk 1 above)
Comment (lpetrovic05) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/GossipEvent.java:390]: yeah, i agree

Comment (lpetrovic05) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/GossipEvent.java:390]: in fact, we do use these for production. which is why I think I should change this method. we put them in hash sets for production. this is something I need to fix

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/system/events/DetailedConsensusEvent.java (hunk 5 above)
Comment (lpetrovic05) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/system/events/DetailedConsensusEvent.java:83]: done

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/system/events/DetailedConsensusEvent.java (hunk 6 above)
Comment (lpetrovic05) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/system/events/DetailedConsensusEvent.java:91]: added

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/system/events/DetailedConsensusEvent.java (hunk 7 above)
Comment (lpetrovic05) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/system/events/DetailedConsensusEvent.java:133]: it does not, there is a unit test that checks it

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/EventUtils.java (hunk 10 above)
Comment (lpetrovic05) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/EventUtils.java:63]: added

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/EventUtils.java (hunk 11 above)
Comment (lpetrovic05) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/EventUtils.java:71]: added

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/EventUtils.java (hunk 12 above)
Comment (lpetrovic05) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/EventUtils.java:82]: added

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/internal/EventImpl.java (hunk 13 above)
Comment (lpetrovic05) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/internal/EventImpl.java:103]: added

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/system/events/DetailedConsensusEvent.java (hunk 14 above)
Comment (lpetrovic05) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/system/events/DetailedConsensusEvent.java:95]: i added a comment with an explenation

**Review by alittley — DISMISSED** [review]

**Review by alittley — APPROVED** [review]

**Review by jeromy-cannon — APPROVED** [review]

**Pull request merged**